apollo-router/src/test_harness/http_snapshot.rs

```diff
@@ -0,0 +1,563 @@
+        error_message = message.clone();
+        error!(
+            url = %uri,
+            method = %method,
+            message,
+        );
+    } else {
+        info!(
+            url = %uri,
+            method = %method,
+            "Taking snapshot"
+        );
+        let mut request = http::Request::builder()
+            .method(method.clone())
+            .version(version)
+            .uri(uri.clone())
+            .body(router_body)
+            .unwrap();
+        *request.headers_mut() = request_headers.clone();
+        let response = state
+            .client
+            .oneshot(HttpRequest {
+                http_request: request,
+                context: crate::context::Context::new(),
+            })
+            .await
+            .unwrap();
+        let (parts, body) = response.http_response.into_parts();
+
+        if let Ok(body_bytes) = body.to_bytes().await {
+            if let Ok(response_json_body) = serde_json::from_slice(&body_bytes) {
+                let snapshot = Snapshot {
+                    request: Request {
+                        method: Some(method.to_string()),
+                        path: Some(path),
+                        body: request_json_body,
+                    },
+                    response: Response {
+                        status: parts.status.as_u16(),
+                        headers: map_headers(parts.headers, |name| {
+                            state
+                                .include_headers
+                                .as_ref()
+                                .map(|headers| headers.contains(&name.to_string()))
+                                .unwrap_or(true)
+                        }),
+                        body: response_json_body,
+                    },
+                };
+                {
+                    let mut snapshots = state.snapshots.lock().unwrap();
+                    snapshots.insert(snapshot.key(), snapshot.clone());
+                    if let Err(e) = save(state.snapshot_file, &mut snapshots) {
+                        error!(
+                            url = %uri,
+                            method = %method,
+                            error = ?e,
+                            "Unable to save snapshot"
+                        );
+                    }
+                }
+                return Ok(snapshot.try_into().unwrap());
+            } else {
+                let message = "Unable to parse response body as JSON".to_string();
+                error_message = message.clone();
+                error!(
+                    url = %uri,
+                    method = %method,
+                    message,
+                )
+            }
+        } else {
+            let message = "Unable to read response body".to_string();
+            error_message = message.clone();
+            error!(
+                url = %uri,
+                method = %method,
+                message
+            )
+        }
```

Comment:
Small nit: this is a bit gnarly. It might be nice to break this up into fallible sections that use the `?` to return early, but up to you.

Reply:
More than fair. I'll take a look at cleaning this up.